torch/csrc/dynamo/eval_frame.c

```diff
@@ -543,16 +564,22 @@ static PyObject* lookup(CacheEntry* e, THP_EVAL_API_FRAME_OBJECT *frame, CacheEn
         e->next = extra;
         set_extra(frame->f_code, e);
     }
+    Py_DECREF(result);
     return (PyObject*)e->code;
   }
-  if (unlikely(guard_fail_hook != NULL)) {
-    PyObject* r = call_guard_fail_hook(guard_fail_hook, e, index, f_locals);
-    if (r == NULL) {
-      return NULL;
-    }
-    Py_DECREF(r);
+  // valid == False
+  // fail_installed_guard_subexpression is a borrowed ref here
+  PyObject* fail_installed_guard_subexpression = PyTuple_GetItem(result, 1);
+  Py_DECREF(result);
+  PyObject* lookup_result = lookup(e->next, frame, e, index + 1, installed_guard_subexpression);
```

Comment:
```
def f(x, y):
  if x:  # x: bool
    return y + 2
  else:
    return y * 2
```

f(False, torch.randn(3))
x == False and y.size(0) == 3

{y.size(0): 3}

f(True, torch.randn(3))
x == True and y.size(0) == 3

same

f(False, torch.randn(4))

{y.size(0): None}